hypershift-operator/controllers/hostedClusterInfrastructure/infra_interface.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hostedclusterinfrastructure
+
+import (
+	"context"
+	"errors"
+
+	corev1 "k8s.io/api/core/v1"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	awscore "github.com/aws/aws-sdk-go/aws"
+	"github.com/go-logr/logr"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/cmd/infra/aws"
+	"github.com/openshift/hypershift/cmd/infra/azure"
+)
+
+const (
+	awsAccessKey = "aws_access_key_id"
+	awsSecretKey = "aws_secret_access_key"
+)
+
+type InfraHandler interface {
+	AwsInfraCreator(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AwsCreateInfra
+	AwsInfraDestroyer(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AwsDestroyInfra
+	AwsIAMCreator(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AwsCreateIAM
+	AwsIAMDestroyer(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AwsDestroyIAM
+
+	AzureInfraDestroyer(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AzureDestroyInfra
+	AzureInfraCreator(hci *hyperv1.HostedClusterInfrastructure, providerSecret *corev1.Secret) AzureCreateInfra
+}
```

Comment:
It seems like a generic interface would be better here. Something that isn't provider specific, or for combined use of multiple providers.

Reply:
This was done, to allow for FVT and not specifically to hide the types.

I made the following changes based on this ask and the following ask.
1. I moved the create/destroy to a switch to deal with the three supported providers, stubbed out messages for the two that still need support.
2. I moved most of the AWS specific calls to the aws_infra.go
3. I left these interfaces as they were done to stub out the actual AWS codes for FVT tests

Reply:
I further generalized the flow and inserted some messaging for Azure and PowerVS.